src/Toml/Type.hs

```diff
@@ -0,0 +1,146 @@
+{- | Contains specification of TOML via Haskell ADT. -}
+
+module Toml.Type
+       ( TOML (..)
+       , Key (..)
+       , TableId (..)
+       , Value (..)
+       , DateTime (..)
+       ) where
+
+import Data.Hashable (Hashable)
+import Data.HashMap.Strict (HashMap)
+import Data.List.NonEmpty (NonEmpty)
+import Data.Text (Text)
+import Data.Time (Day, LocalTime, TimeOfDay, ZonedTime)
+
+{- | Key of value in @key = val@ pair.
+-}
+newtype Key = Key
+    { unKey :: Text
+    } deriving (Eq, Ord, Hashable)
+
+{- | Name of table. Stored as 'NonEmpty' list of components.
+
+@
+[table.name]
+  key1 = "val1"
+  key2 = "val2"
+@
+-}
+newtype TableId = TableId
+    { unTableId :: NonEmpty Text
+    } deriving (Eq, Ord, Hashable)
+
+-- TODO: describe how some TOML document will look like with this type
+{- | Represents TOML configuration value. -}
+data TOML = TOML
+    { tomlPairs       :: HashMap Key     Value
+    , tomlTables      :: HashMap TableId TOML
+    , tomlTableArrays :: HashMap TableId (NonEmpty TOML)
+    -- TODO: I don't really like the above structure, probably something better should be used
```

Comment:
That's good point, but I don't have anything in mind for the moment..

Reply:
The first version of parser won't support array of tables anyways, so we can safely comment this line. And maybe we will come up with better design later...